tests/backends/test_base.py

```diff
@@ -0,0 +1,87 @@
+import importlib
+
+import pytest
+
+from cohortextractor.backends.base import BaseBackend, Column, MappedTable
+from cohortextractor.concepts import types
+from cohortextractor.concepts.table_contract import BackendContractError
+from cohortextractor.concepts.table_contract import Column as ColumnContract
+from cohortextractor.concepts.table_contract import TableContract
+from cohortextractor.query_engines.base_sql import BaseSQLQueryEngine
+
+
+class PatientsContract(TableContract):
+    """
+    A test contract that backends can be validates against
+    """
+
+    patient_id = ColumnContract(type=types.PseudoPatientId(), help="")
+    date_of_birth = ColumnContract(type=types.Date(), help="")
+    sex = ColumnContract(type=types.Choice("F", "M"), help="")
+
+
+def test_bad_backend_with_validate_contract():
+
+    # Create a Backend which will fail to implement a contract
+    class BadBackend(BaseBackend):
+        backend_id = "bad_test_backend"
+        query_engine_class = BaseSQLQueryEngine
+        patient_join_column = "patient_id"
+
+        patients = MappedTable(
+            implements=PatientsContract,
+            source="Patient",
+            columns=dict(
+                date_of_birth=Column("date", source="DateOfBirth"),
+            ),
+        )
+
+    bad_backend = BadBackend(database_url="test")
+
+    with pytest.raises(BackendContractError, match="Missing columns: sex"):
+        bad_backend.validate_all_contracts()
+
+
+def test_good_backend_with_validate_contract():
+    # Create a Backend which will implement a contract
+    class GoodBackend(BaseBackend):
+        backend_id = "good_test_backend"
+        query_engine_class = BaseSQLQueryEngine
+        patient_join_column = "patient_id"
+
+        patients = MappedTable(
+            implements=PatientsContract,
+            source="Patient",
+            columns=dict(
+                date_of_birth=Column("date", source="DateOfBirth"),
+                sex=Column("varchar", source="Sex"),
+            ),
+        )
+
+    good_backend = GoodBackend(database_url="test")
+    good_backend.validate_all_contracts()
+
+
+def test_validate_all_backends():
+    """
+    Loops through all the backends, excluding test ones,
+    and validates they meet any contract that they claim to
+    """
+    real_backends = [cls.__name__ for cls in BaseBackend.__subclasses__()]
+    dummy_backends = ["BadBackend", "GoodBackend"]
+
+    for backend in real_backends:
+        # Exclude test backends
+        if backend not in dummy_backends:
+
+            # get the class from cohortextractor.backends
+            backend_class = getattr(
+                importlib.import_module(
+                    f"cohortextractor.backends.{backend[:-7].lower()}"
+                ),
+                backend,
+            )
+
+            # Instantiate the class
+            backend_instance = backend_class(database_url="test")
+            backend_instance.validate_all_contracts()
```

Comment:
`validate_all_contracts()` is a class method, so you don't need to create an instance here, you can call it directly on the class.

Reply:
good point - remvoed.